homeassistant/components/switchbot_cloud/fan.py

```diff
@@ -0,0 +1,114 @@
+"""Support for the Switchbot Battery Circulator fan."""
+
+from typing import Any
+
+from switchbot_api import (
+    BatteryCirculatorFanCommands,
+    BatteryCirculatorFanMode,
+    CommonCommands,
+)
+
+from homeassistant.components.fan import FanEntity, FanEntityFeature
+from homeassistant.config_entries import ConfigEntry
+from homeassistant.core import HomeAssistant
+from homeassistant.helpers.entity_platform import AddConfigEntryEntitiesCallback
+
+from . import SwitchbotCloudData
+from .const import DOMAIN
+from .entity import SwitchBotCloudEntity
+
+
+async def async_setup_entry(
+    hass: HomeAssistant,
+    config: ConfigEntry,
+    async_add_entities: AddConfigEntryEntitiesCallback,
+) -> None:
+    """Set up SwitchBot Cloud entry."""
+    data: SwitchbotCloudData = hass.data[DOMAIN][config.entry_id]
+    async_add_entities(
+        SwitchBotCloudFan(data.api, device, coordinator)
+        for device, coordinator in data.devices.fans
+    )
+
+
+class SwitchBotCloudFan(SwitchBotCloudEntity, FanEntity):
+    """Representation of a SwitchBot Battery Circulator Fan."""
+
+    _attr_name = None
+
+    _attr_supported_features = (
+        FanEntityFeature.SET_SPEED
+        | FanEntityFeature.PRESET_MODE
+        | FanEntityFeature.TURN_OFF
+        | FanEntityFeature.TURN_ON
+    )
+    _attr_preset_modes = list(BatteryCirculatorFanMode)
+
+    _attr_is_on: bool | None = None
+
+    @property
+    def is_on(self) -> bool | None:
+        """Return true if the entity is on."""
+        return self._attr_is_on
+
+    def _set_attributes(self) -> None:
+        """Set attributes from coordinator data."""
+        if self.coordinator.data is None:
+            return
+
+        power: str | None = self.coordinator.data.get("power")
+        mode: str | None = self.coordinator.data.get("mode")
+        fan_speed: str | None = self.coordinator.data.get("fanSpeed")
+        assert power is not None
+        self._attr_is_on = "on" in power
+        assert mode is not None
+        self.preset_mode = mode
+        assert fan_speed is not None
+        self.percentage = int(fan_speed)
+
+    async def async_turn_on(
+        self,
+        percentage: int | None = None,
+        preset_mode: str | None = None,
+        **kwargs: Any,
+    ) -> None:
+        """Turn on the fan."""
+        await self.coordinator.async_refresh()
+        await self.send_api_command(CommonCommands.ON)
+        self._attr_is_on = True
+        self.async_write_ha_state()
```

Comment:
But what if someone turns it on with a percentage and preset mode?

You by the way now refresh the coordinator which is great, but shouldn't we do that at the end? Then we don't write state in here, but just write state with the newly gathered data

Reply:
> But what if someone turns it on with a percentage and preset mode?
> 
> You by the way now refresh the coordinator which is great, but shouldn't we do that at the end? Then we don't write state in here, but just write state with the newly gathered data
1. added new code for  **turns it on with percentage and preset mode**
2. I do not want write state after refresh,   but it always show wrong state, when I test it with real device , maybe cause by bad service interface